### ISSUE TITLE
scipy.optimize has broken my trust.

### ISSUE THREAD

**Issue by brandondube** [scipy/scipy#8373]
This submodule has broken my trust.  

I am genuinely in awe of how it was included in scipy 1.0.  I am tagging two heavy hitters to attempt to bring discussion to this issue since, in general, scipy.optimize issues seem to go without discussion or fixes.  @pv @teoliphant

As I understand, an effort was made in the last few version of scipy leading up to and including 1.0 to unify the interface of the different optimization routines and eventually deprecate fmin_xxxx.

A fairly large issue with this is that many of the inputs and outputs in the scipy documentation for these functions are simply wrong.

Take, for example the [BFGS method](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-bfgs.html#optimize-minimize-bfgs).

the documentation for disp is incorrect; convergence messages are not printed, only a blurb at the end.  return_all appears in the call signature at the top and is undocumented.  I looked to [fmin_bfgs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.fmin_bfgs.html) hoping its more specific documentation would be superior.  The documentation for that is even more misleading.  I hoped that the specified type for the constituents of allvecs being OptimizationResult would mean they hold cost function values and input vector values.  They are simply input vector values.  The documentation here is also misleading.

I also understand that many of the algorithms are wrapped code in another language, e.g. L-BFGS-B is written in FORTRAN77, and this may make their modification difficult.  Earlier today I opened an issue that L-BFGS-B does not respect the maxiter input, #8372.  There are other issues, e.g. lack of respect for user's input dtype because of hard coded doubles in the fortran code, for example.

However, the docs being misleading is a great travesty for the users of this module, which I believe is among the more important/widely used of scipy.  I would work to improve them myself, but I do not feel that my knowledge of the inner workings of the module is great enough to do so without myself making mistakes.

Returning to L-BFGS-B, the wrapped fortran code appears to contain a variety of errors or behaviors that have a varied level of harmfulness.  For example, this is an example output:

```
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =            4     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  2.75370D+01    |proj g|=  1.06893D+02

At iterate    1    f=  2.27149D+01    |proj g|=  1.76623D+02

At iterate    2    f=  1.20196D+01    |proj g|=  1.20733D+02

At iterate    3    f=  8.05081D+00    |proj g|=  8.74848D+01

At iterate    4    f=  5.00344D+00    |proj g|=  8.42253D+01

At iterate    5    f=  3.70940D+00    |proj g|=  1.01991D+02

At iterate    6    f=  1.57818D+00    |proj g|=  7.10948D+01

At iterate    7    f=  4.81956D-01    |proj g|=  3.76945D+01

At iterate    8    f=  6.81301D-02    |proj g|=  2.20645D+01

At iterate    9    f=  1.34639D-02    |proj g|=  8.79078D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
    4      9     12      1     0     0   8.791D+00   1.346D-02
  F =  1.346394518877836E-002

CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH             

 Cauchy                time 0.000E+00 seconds.
 Subspace minimization time 0.000E+00 seconds.
 Line search           time 0.000E+00 seconds.

 Total User time 0.000E+00 seconds.
```
the code writes "iterate" instead of "iteration" (a typo).  This is an annoyance as it is not very professional, but it does not break anything.  The separate timings for Cauchy, subspace minimization, and line search are also always zero.  Same for user time.  It would be nice for these to not be a filler values since this is misleading, but same thing - this is creature comfort problem, not an accuracy problem.

My application is in a topic known as wavefront sensing, or [phase retrieval](https://en.wikipedia.org/wiki/Phase_retrieval).  We put in some number of parameters, and the landscape of the cost function is highly evolutionary as the optimizer traverses the space.  As a result, when designing a cost function or debugging code it is desirable to know the state of the optimizer, even without knowing its internal mechanisms, as it walks the space.  To do this I do the following:

1.  Save this output text from L-BFGS-B, and scrape the f= values from it.
2.  Use a callback that stores parameter vectors.

I expect these two things to have the same length always, and depend on cost[n] being associated with all_x[n] without registration errors between these two blocks of data.

In my particular application (I can share code on request, but 1000-2000 lines are called to implement my routine split across a few modules; there is no easy "def foo()" analog that I have found to reproduce this) I am having trouble understanding why my cost function appears to not be correlated, based on the output of l-bfgs-b, to the closeness of the current iteration input to truth:

![cost function is not correlated with true error metric](https://i.imgur.com/Br9OHCg.jpg)

This has cost me dozens of hours of debugging time, only to learn that **something is wrong with L-BFGS-B that causes the output of options={'disp': True} to not be the the cost function at a given parameter vector.**

Because I capture stdout, I am not able to use print.  I have embedded a line in my cost function that appends the current parameter vector and cost to a .txt file and resorted to matching values in that to L-BFGS-B print output to see what is wrong.  The text from that is below, with numbered lines

```
00. params = [ 0.  0.  0.  0.] 	 f(x) = 27.5369873872

01. params = [  1.00000000e-08   0.00000000e+00   0.00000000e+00   0.00000000e+00] 	 f(x) = 27.5369884562

02. params = [  0.00000000e+00   1.00000000e-08   0.00000000e+00   0.00000000e+00] 	 f(x) = 27.5369867324

03. params = [  0.00000000e+00   0.00000000e+00   1.00000000e-08   0.00000000e+00] 	 f(x) = 27.5369870096

04. params = [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e-08] 	 f(x) = 27.5369875808

05. params = [-0.80769268  0.49476373  0.28537275 -0.14630029] 	 f(x) = 86.4418490547

06. params = [-0.80769267  0.49476373  0.28537275 -0.14630029] 	 f(x) = 86.4418490135

07. params = [-0.80769268  0.49476374  0.28537275 -0.14630029] 	 f(x) = 86.441849366

08. params = [-0.80769268  0.49476373  0.28537276 -0.14630029] 	 f(x) = 86.4418499781

09. params = [-0.80769268  0.49476373  0.28537275 -0.14630028] 	 f(x) = 86.4418482602

10. params = [-0.16406864  0.1005026   0.05796848 -0.02971834] 	 f(x) = 22.714940612

11. params = [-0.16406863  0.1005026   0.05796848 -0.02971834] 	 f(x) = 22.7149388457

12. params = [-0.16406864  0.10050261  0.05796848 -0.02971834] 	 f(x) = 22.7149405671

13. params = [-0.16406864  0.1005026   0.05796849 -0.02971834] 	 f(x) = 22.7149404753

14. params = [-0.16406864  0.1005026   0.05796848 -0.02971833] 	 f(x) = 22.7149421094

15. params = [-0.10087717  0.09158391  0.05575493 -0.06632038] 	 f(x) = 12.0195875306

16. params = [-0.10087716  0.09158391  0.05575493 -0.06632038] 	 f(x) = 12.0195863232

17. params = [-0.10087717  0.09158392  0.05575493 -0.06632038] 	 f(x) = 12.0195878653

18. params = [-0.10087717  0.09158391  0.05575494 -0.06632038] 	 f(x) = 12.0195883001

19. params = [-0.10087717  0.09158391  0.05575493 -0.06632037] 	 f(x) = 12.0195867159

20. params = [ 0.15842029 -0.01548538 -0.01632799 -0.07159999] 	 f(x) = 48.7174210299

21. params = [ 0.1584203  -0.01548538 -0.01632799 -0.07159999] 	 f(x) = 48.7174233933

22. params = [ 0.15842029 -0.01548537 -0.01632799 -0.07159999] 	 f(x) = 48.717421743

23. params = [ 0.15842029 -0.01548538 -0.01632798 -0.07159999] 	 f(x) = 48.7174193993

24. params = [ 0.15842029 -0.01548538 -0.01632799 -0.07159998] 	 f(x) = 48.7174245544

25. params = [-0.05512618  0.07269238  0.04303647 -0.06725192] 	 f(x) = 8.05081234605

26. params = [-0.05512617  0.07269238  0.04303647 -0.06725192] 	 f(x) = 8.05081192026

27. params = [-0.05512618  0.07269239  0.04303647 -0.06725192] 	 f(x) = 8.0508114712

28. params = [-0.05512618  0.07269238  0.04303648 -0.06725192] 	 f(x) = 8.05081281076

29. params = [-0.05512618  0.07269238  0.04303647 -0.06725191] 	 f(x) = 8.05081206742

30. params = [-0.02149653  0.09172596  0.0267816  -0.06652566] 	 f(x) = 5.0034401551

31. params = [-0.02149652  0.09172596  0.0267816  -0.06652566] 	 f(x) = 5.00344002938

32. params = [-0.02149653  0.09172597  0.0267816  -0.06652566] 	 f(x) = 5.00343999403

33. params = [-0.02149653  0.09172596  0.02678161 -0.06652566] 	 f(x) = 5.00344079846

34. params = [-0.02149653  0.09172596  0.0267816  -0.06652565] 	 f(x) = 5.00343931285

35. params = [ 0.03012534  0.12108955 -0.01347029 -0.04719621] 	 f(x) = 3.70939759272

36. params = [ 0.03012535  0.12108955 -0.01347029 -0.04719621] 	 f(x) = 3.70939819132

37. params = [ 0.03012534  0.12108956 -0.01347029 -0.04719621] 	 f(x) = 3.70939842652

38. params = [ 0.03012534  0.12108955 -0.01347028 -0.04719621] 	 f(x) = 3.70939740919

39. params = [ 0.03012534  0.12108955 -0.01347029 -0.0471962 ] 	 f(x) = 3.70939657281

40. params = [ 0.00300894  0.11308784  0.00337076 -0.03858502] 	 f(x) = 1.57818376805

41. params = [ 0.00300895  0.11308784  0.00337076 -0.03858502] 	 f(x) = 1.578184007

42. params = [ 0.00300894  0.11308785  0.00337076 -0.03858502] 	 f(x) = 1.57818379941

43. params = [ 0.00300894  0.11308784  0.00337077 -0.03858502] 	 f(x) = 1.57818414645

44. params = [ 0.00300894  0.11308784  0.00337076 -0.03858501] 	 f(x) = 1.5781830571

45. params = [-0.01561316  0.12143126  0.00100096 -0.01330919] 	 f(x) = 0.481955650202

46. params = [-0.01561315  0.12143126  0.00100096 -0.01330919] 	 f(x) = 0.481955273258

47. params = [-0.01561316  0.12143127  0.00100096 -0.01330919] 	 f(x) = 0.481955840221

48. params = [-0.01561316  0.12143126  0.00100097 -0.01330919] 	 f(x) = 0.481955839908

49. params = [-0.01561316  0.12143126  0.00100096 -0.01330918] 	 f(x) = 0.481955373719

50. params = [-0.00041042  0.12401147 -0.00485423  0.00207093] 	 f(x) = 0.0681301442419

51. params = [-0.00041041  0.12401147 -0.00485423  0.00207093] 	 f(x) = 0.0681301690496

52. params = [-0.00041042  0.12401148 -0.00485423  0.00207093] 	 f(x) = 0.0681301089522

53. params = [-0.00041042  0.12401147 -0.00485422  0.00207093] 	 f(x) = 0.0681299235971

54. params = [-0.00041042  0.12401147 -0.00485423  0.00207094] 	 f(x) = 0.0681302701249

55. params = [-0.00144057  0.12278377 -0.00124594 -0.00264166] 	 f(x) = 0.0134639451888

56. params = [-0.00144056  0.12278377 -0.00124594 -0.00264166] 	 f(x) = 0.0134639411113

57. params = [-0.00144057  0.12278378 -0.00124594 -0.00264166] 	 f(x) = 0.0134638572809

58. params = [-0.00144057  0.12278377 -0.00124593 -0.00264166] 	 f(x) = 0.0134639506651

59. params = [-0.00144057  0.12278377 -0.00124594 -0.00264165] 	 f(x) = 0.0134639172132
```

The callback creates the following list:
```
[array([ 0.,  0.,  0.,  0.]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166]),
  array([-0.00144057,  0.12278377, -0.00124594, -0.00264166])]
```
I have reproduced this for completeness, but the important lines are 00, the initial guess, which matches iterate 0 in the output text from the optimizer.  Iteration 1 matches any of line 12-14.  The parameter vectors are similar to:
```
[-0.16406864  0.10050261  0.05796848 -0.02971834]
```
The callback contains this input array at that iteration:
```
[-0.00144057,  0.12278377, -0.00124594, -0.00264166]
```
Which would be associated with a cost function 0.013, not 22 and change.  As a result, I am unable to trust the tools I have to diagnose a problem with them.  The output of the L-BFGS-B print simply is not the value of the cost function for the parameter vector at that iteration.  I believe the problem is something like as follows:

1.  The callback is called at the end of a line search before beginning a new gradient hunt
2.  The f= is written at the start of a line search

But I cannot trust the documentation of scipy to be correct that the callback is called at the end of an iteration (~= end of line search in BFGS), and I cannot trust lbfgsb.f to be error free in this way.

Are there plans to improve the optimization module in a "months not years" timespan?  Is anyone willing to rewrite lbfgsb.f or even port it to C/C++ or pure python to solve these problems?  In commercialized applications, it is often desirable for a user to keep their finger on the pulse by monitoring convergence and potentially stopping optimization, hand tuning the inputs, and then continuing the work.  The architecture of optimize.minimize is prohibitive to this.  It would be nice to see a context manager based implementation that allows interruption of the optimizer without losing the current state, as this sort of behavior is only possible at the moment through use of a complicated callback function.

**Comment by andyfaff** [comment]
There are many issues to unpack here.

>  There are other issues, e.g. lack of respect for user's input dtype because of hard coded doubles in the fortran code, for example.

double precision is needed internally when calculating gradients because numerical inaccuracies are created in the gradient otherwise.

> the code writes "iterate" instead of "iteration" (a typo). This is an annoyance as it is not very professional, but it does not break anything. 

I can assure you that the whole scipy community works very hard to develop this software, in their own free time, doing their best. To me your post comes across as unprofessional. For future information, `iterate` is well used in this context.
```
noun: iterate; plural noun: iterates
1.
a quantity arrived at by iteration.
```

There is an issue with maxiter, reported in #7854, which does need to be corrected, this will eventually be tackled, along with the ~1000 other issues that are destined to be addressed.

> As a result, when designing a cost function or debugging code it is desirable to know the state of the optimizer, even without knowing its internal mechanisms, as it walks the space. To do this I do the following: Save this output text from L-BFGS-B, and scrape the f= values from it. Use a callback that stores parameter vectors.

Scraping the output for this purpose is not a robust way to do this. Ideally the callback function should be supplied an `OptimizeResult` which contains the best `x` so far, and the function evaluation. This would be a great new feature, but has to take into account back compatibility. In the meantime you should be able to calculate f(x) in the callback function, or even memoize f(x) for given `x`.

> This has cost me dozens of hours of debugging time, only to learn that something is wrong with L-BFGS-B that causes the output of options={'disp': True} to not be the the cost function at a given parameter vector.

I'm sorry that you lost dozens of hours on this. Scraping output is not a robust way of doing this.

> Are there plans to improve the optimization module in a "months not years" timespan? Is anyone willing to rewrite lbfgsb.f or even port it to C/C++ or pure python to solve these problems?

We continually look for improvements, but this is mostly done in peoples free time. If you are willing to rewrite the codebase, we would welcome a Pull Request.

> It would be nice to see a context manager based implementation that allows interruption of the optimizer without losing the current state, as this sort of behavior is only possible at the moment through use of a complicated callback function.

I have plans to implement this for one solver, but gradual evolution is required, that would require a lot of work for all the solvers, and one would want a fairly uniform approach across them all.

**Comment by andyfaff** [comment]
Having had a look at lbfgsb.f, it's highly unlikely that'll be ported to another language anytime soon. In fact there's a benefit to keeping it in fortran - we get to use any updated versions that are [released](http://users.iems.northwestern.edu/~nocedal/lbfgsb.html).

**Comment by josef-pkt** [comment]
> This submodule has broken my trust.

I would never trust any low level implementation details unless they are checked and verified by unit tests. On the higher level, I spent weeks over the years fighting with convergence problems across optimizers and scipy version, e.g. a few years ago I wrote this https://github.com/statsmodels/statsmodels/issues/1641 , but I never worried about whether maxiter is defined as number of iterations or number of iterations plus one except in the case of maxiter=0 versus maxiter=1.

Overall I worry that the "feature creep" in optimize introduces a large amount of overhead and makes it more difficult for tools like numba to speed it up. (disclaimer: I don't have much experience with  numba yet.)

Aside: I'm overall pretty happy with scipy optimize improvements and I especially like that development is going into getting new, better or more robust optimizers, (speaking as statsmodels convergence problem hunter).

**Comment by brandondube** [comment]
> double precision is needed internally when calculating gradients because numerical inaccuracies are created in the gradient otherwise.

This depends on the value of epsilon and the overall scale of the cost function and I do not believe can be stated as fact.  If a user provides f32s, the library may warn the user but should respect their wishes choice of data type.

> To me your post comes across as unprofessional.

Then it is a good thing no one here is acting in a professional role.  I am highly frustrated that an authoritative library like scipy has a submodule which is wrought with errors that are nonobvious and bite you very hard when they appear.  This issue is an expression of that deep frustration.

> For future information, iterate is well used in this context.

I would defer to [english.se](https://english.stackexchange.com/a/41749) that it is not.

> There is an issue with maxiter, reported in #7854, which does need to be corrected, this will eventually be tackled, along with the ~1000 other issues that are destined to be addressed.

There are << 1000 issues specific to the optimization submodule.

> Scraping the output for this purpose is not a robust way to do this. 

If the output can be trusted, it is a perfectly fine way to do it.  The textual output is machine generated and you can expect scraping it to not silently fail.  My cost function is expensive - about 1s/evaluation - so I would prefer to not play back, say, 150 iterations after it has completed to compute the cost function trajectory.  This takes < 1ms and is, effectively, free.

> Ideally the callback function should be supplied an OptimizeResult which contains the best x so far, and the function evaluation. This would be a great new feature, but has to take into account back compatibility.

The mere mention of OptimizeResult worries me, since this is referenced over and over again in the documentation for optimize.minimize and optimize.fmin_xxxx but is rarely actually used.  The bad taste is so strong cleansing that name from scipy.optimizes vocabulary would be preferred if the idea were to be consistently used and documented.

[Two months ago](https://github.com/scipy/scipy/issues/8176) I opened an issue about this very thing.  I do not think it was a bad design choice to unify the solver interfaces under optimize.minimize, *but they are not truly uniform because of the options dict.*  To be pythonic, that should have been **kwargs, but whatever.  Introducing another kwarg/option that changes what is fed to the callback would:

1.  I think not confuse users
2.  Be a decently elegant solution

It would look something like
```python
# in minimize.py::minimize
if callback is not None:
    if cost_all is True:
        callback(x, fx)
    else:
        callback(x)
```
Very simple.  If there is concern about this breaking compatibility, just make scipy 2.0 with breaking changes, or scipy 1.1 if 0.x versions can make breaking changes.  This change wouldn't be breaking, though.

> or even memoize f(x) for given x.

f(x) is passed numpy arrays.  Memoization is nontrivial in that scenario.  You can do `np.all(a==b)` but FP roundoff might bite you.  There would be an art in using `np.allclose` to not break the optimizer itself by accidentally returning the wrong cost function value for some small epsilon.

> We continually look for improvements, but this is mostly done in peoples free time.

Scipy has long reached critical mass where it is an authoritative package.  Free time or not, version 1.0 signifies production ready for real, and on pypi scipy is listed as development 5, production/stable.  optimize.minimize does not fit that criteria, and should not even be included in its current state.  You may disagree, but as a _user_ with no stake in Scipy other than my familiarity with it and no malice towards what is on the whole a great effort, that is my opinion and I am entitled to it.

> If you are willing to rewrite the codebase

I do not have a py2.7 install, and would not want to install one to develop on.  All the tricks like importing six and other black magic to make a module work with 2.7 and 3.x are not things I want to learn.  Of greater importance is that I am not an expert in nonlinear optimization and there would be significant nonzero risk in me making a bad error.

**Comment by xoviat** [comment]
No one is forcing you to use SciPy. If you want commercial support, then you can pay for a MATLAB license. 

> You may disagree, but as a user with no stake in Scipy other than my familiarity with it and no malice towards what is on the whole a great effort, that is my opinion and I am entitled to it.

You fundamentally misunderstand open source software. Fundamentally. SciPy and other open source software isn't written for you. It's written for its contributors, and you're allowed to use it as a side effect. If you're not a contributor, then you are not entitled to decide what it should be called, its development status, or any other aspect of its development or distribution. Perhaps you can convince a contributor to adopt your opinion, but that's unlikely with the tone that you display here.

**Comment by brandondube** [comment]
> You fundamentally misunderstand open source software. Fundamentally.

No, I do not.  I write a few open source packages myself - some on github, some not - and on several occasions have changed the interface to something, or its internal workings, based on user feedback.


You fundamentally misunderstand human interaction.  Fundamentally.

**Comment by xoviat** [comment]
This issue isn't a bug report. It's an attempt to claim the time of others to work on issues that they themselves have decided not to prioritize:

>  I am tagging two heavy hitters to attempt to bring discussion to this issue since, in general, scipy.optimize issues seem to go without discussion or fixes. @pv @teoliphant

**Comment by brandondube** [comment]
@xoviat please go away.

**Comment by ilayn** [comment]
I think this is not a healthy way to discuss. The license of SciPy clearly mentions the very problem you are addressing. 

We are sorry for your experience but we are not obliged to take immediate action to remedy this because an open source library does not operate on the terms you would be expecting as such from a commercial software.

I would really recommend not discussing this further unless there is a code contribution or an isolated issue.

**Comment by brandondube** [comment]
@ilayn I do not think asking for "months not years" is asking for immediate action.

Issues related to scipy.optimize have been largely ignored on this repository.  I think this is a very major problem with optimize.minimize, or at least with method='L-BFGS-B', and think it needs to be addressed.  I pinged two of the biggest names re: scipy to draw attention to this and gave it a dramatic name.  So far, it appears to have worked, but not in the way that I would have wanted.

**Comment by ilayn** [comment]
It might be a major problem for you. Every part of SciPy has different problems big or small. But they are not fixed on demands or frustrations. I have worked on Pull Requests that took more than a year. I do still wait for action on some items. Of course they have to be addressed but definitely not by demanding and complaining about a lot of people's voluntary time and patience. 

Those biggest names are not the cavalary you hope them to arrive. You seem to have a point of view that doesn't agree with the reality of how SciPy works.

**Comment by WarrenWeckesser** [comment]
@xoviat wrote

> SciPy and other open source software isn't written for you. It's written for its contributors, and you're allowed to use it as a side effect.

Please be assured that this is *not* the opinion of all (or even most, I suspect) SciPy developers.  The library is written for the world to use.  If it were written only for its "contributors", I'd stop contributing.

Moreover, by filing this issue, Brandon *is* contributing.  It took a lot of time to create that initial report; he could have simply stopped using it SciPy.  Feedback--good or bad--is essential, and should not be ignored.  Some of the comments may lean more towards "rant" than "bug report", but try to put yourself in his shoes when he says "This has cost me dozens of hours of debugging time".   You'd probably be ranting too.  Please try to filter out the emotional aspects of the report ("great travesty", etc.) and focus on the defects--big and small--that are being reported.

**Comment by ilayn** [comment]
As I said unless there is an isolated issue or code contribution my recommendation is to stop that kind of discussion. Just to state in case it is not clear, I am not trying to close this issue but it has to be streamlined and cooled down a little bit.

**Comment by tempuser-111** [comment]
For some reason, I am no longer able to comment with my standard account on this issue (I was personally unaware that it was possible to block a specific user account on a specific thread without blocking the user from the entire repository, however it's possible that one of my comments was flagged).

However, I wanted to note that there was no rational basis to post my original comment and therefore, it must have been made in err.

- @xoviat

**Comment by brandondube** [comment]
I have found that the discrepancy between the callback and L-BFGS-B output will manifest with the rosen function.

An example:
```python
from scipy.optimize import minimize, rosen
# import forcefully_redirect_stdout and parse_cost_by_iter_lbfgsb

params = []
def callback(x):
    params.append(x)


with forcefully_redirect_stdout() as out:
    res = minimize(rosen, (0,0,0), options={'disp': True}, method='L-BFGS-B', callback=callback)

costiter = parse_cost_by_iter_lbfgsb(out['txt'])
costiter2 = [rosen(x) for x in params]
print(costiter)
print(costiter2)
```
You will need the two functions from here, https://gist.github.com/brandondube/2c319727bbc794e97c0806a3abd213ba

The output is:
```
[2.0, 1.95917, 1.61222, 1.5741, 1.4304, 1.2286, 0.689151, 0.63911, 0.438557, 0.292715, 0.23319, 0.182176, 0.105509, 0.0824601, 0.0544236, 0.0256008, 0.0108747, 0.00610558, 0.00264869, 0.000595762, 6.15612e-05, 2.10079e-06, 1.12758e-07, 2.97115e-08, 3.8601e-11, 1.58921e-11]
[1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11, 1.5892114699716987e-11]
```
Note that the callback doesn't include the guess (0,0,0), so the second print statement is sheared by one column.  The callback is spammed at the end, one time for each of the times it should have been called along the way.

**Comment by rkern** [comment]
Ah, I think I know the problem. The callback is getting called at the appropriate times, but the `x` array that gets passed into the callback is the same object every time. It gets mutated by the minimizer function on each iteration, so when you finally inspect it at the end, the value is whatever it was at the final iteration. Append a copy of `x` to the list instead.

**Comment by brandondube** [comment]
@rkern  This fixed the problem with the callback being out of sync -- any chance a fix can be made to scipy.optimize.minimize to avoid this behavior?  The parameter arrays are quite small, so I can't imagine making a new one with each tick has particularly big performance implications.

**Comment by rkern** [comment]
Such a fix would be welcomed, certainly; the cost is indeed small, and it would make typical use cases significantly less error-prone. It'll have to be fixed individually in each minimizer implementation, though, sometimes in Python, sometimes in C, depending on the minimizer.

**Comment by andyfaff** [comment]
If the callbacks took an OptimizeResult then the constructor could take care of that.

**Comment by rkern** [comment]
Sure, but that's an even larger change. :-)

**Comment by andyfaff** [comment]
> The callback is getting called at the appropriate times, but the x array that gets passed into the callback is the same object every time. It gets mutated by the minimizer function on each iteration,

I can see both points of view here.

It should be enough to just do something along the lines of (provided there's no performance hit):

```
if callback is not None:
    callback(np.copy(xk))
```

If the `xk` is modified during the function evaluation, then that's another matter :-(.

**Comment by tempuser-112** [comment]
Generally, I don't believe that SciPy should copy the array that is passed to the callback. One of the ways that Python is deceptively simple is that all of its objects are in fact references rather than values. If you were to pass a list into some random function and such a list were to be modified, the list that you had would be modified as well. Similarly, with a NumPy array, you should generally copy the object if you intend to preserve its state.

The behavior here isn't really beginner-friendly, so perhaps a documentation update is in order.

**Comment by andyfaff** [comment]
> Similarly, with a NumPy array, you should generally copy the object if you intend to preserve its state

You're arguing for the status quo here, as your callback is wanting to preserve the state.

**Comment by tempuser-112** [comment]
I am arguing for the status quo, while acknowledging that the status quo could trip-up beginners.

**Comment by tempuser-112** [comment]
Actually not quite-I think we can improve here with a new `full_history` keyword argument that expands the `x`, `fun`, `jac`, and `hess` ndarrays returned to `array.rank + 1`, with the other dimension being the optimization history.

**Comment by brandondube** [comment]
My functions are not mutating `x`, nor is `rosen`.  The mutation is done by L-BFGS-B itself.  You could see a hint to this in that the initial guess was a tuple, which is immutable, at least wholly so in this case where the constituents are ints or floats.

Passing a copy of x to the callback will have negligible performance impact.  Numpy's copy will have a few microseconds of python related overhead, followed by the time to duplicate, in general, a 10 float64 or smaller vector.  In an extreme case, perhaps 1 million parameters which is still 8MB.  Memory moves at around 12GB/s on a modern machine, so ~.7ms to duplicate the array.

In such a scenario the cost function or its gradient is almost assuredly very, very expensive so this overhead is still unimportant.

**Comment by rgommers** [comment]
> For some reason, I am no longer able to comment with my standard account on this issue (I was personally unaware that it was possible to block a specific user account on a specific thread without blocking the user from the entire repository, however it's possible that one of my comments was flagged).

@tempuser-111 I've checked, we have not blocked any users on any repo within the `scipy` organisation. And yes, it's not even possible to block a user on a single issue.

**Comment by tempuser-113** [comment]
@brandondube 

The primary concern is backwards compatibility. If there are callback functions that modify the copy, we would at least need to verify that optimization results are not affected. AFAIK, we cannot have silent breakage.

@rgommers 

The @xoviat, @tempuser-111, and @tempuser-112 have now been blocked from commenting on this issue, and this specific issue only. I'm currently investigating the reason behind these blocks, but this behavior isn't documented, and I don't think that you or any of the administrators here are responsible for it. I have administrative privileges on some repositories, so I know what you can do on the SciPy repository, and I agree with you that you can't block someone from commenting on a particular issue. Honestly I wouldn't bother discussing it further until I identify the cause.

**Comment by tempuser-113** [comment]
Okay, so after some analysis, what appears to have happened is that @brandondube blocked these accounts. My conception of GitHub's logic is currently this:

```python
class Thread(object):
    comments = [Comment(), ...]
    user = User()

class Comment():
   user = User()

if thread.user.name == 'brandontube':
  silentlyBlock()
``` 

It's somewhat worrying that you can block others from commenting on issues, even if they don't own the repository. I'm collecting more data to confirm my suspicions.

**Comment by brandondube** [comment]
@xoviat I asked you to go away.  I will block every account you use to comment on my issue.

**Comment by tempuser-114** [comment]
I don't think I should be blocked from commenting on an issue in the SciPy repository, considering that you don't have privileges here. I also have a script to generate these accounts, so it probably takes more time for you to click the block button.

**Comment by brandondube** [comment]
On the contrary, because I created this issue I have ownership and significant privileges on it.
If you make a 15th alternate account, I will file a report with Github, as you are abusing their services.

**Comment by josef-pkt** [comment]
@rgommers  How can one user block another user?
This is not appropriate behavior for open source.

**Comment by stefanv** [comment]
@brandondube I understand your frustration, both with the library's functionality and Xoviat's comments.  At the same time, it helps to keep in mind that everyone here (including yourself, thank you!) is giving their time freely to debug and improve the library.  When an issue is titled "scipy.optimize has broken my trust", it does not engender a very positive feeling from the developers working on it.  If we steer towards a more technical description such as "scipy.optimize updates parameters in-place" it takes the sting out, and allows us to focus on the issues at hand.

We generally do not prevent others in the community from participating in our comment threads, even when we disagree with their perspectives (perhaps especially then); it's useful to hear those opinions, because one of the strengths of open source is finding a balance between opposing views to build code that is robust for many different use cases.  I think, in this case, you saw the need to protect yourself from Xoviat's comments, which felt personal, and I am sorry for that.  I cannot make you unblock Xoviat, but I do think it would be better if we could all sit around the table and discuss the issues at hand, with an eye towards improving `scipy.optimize`.

Thank you for giving your time to benchmark and to evaluate the code, and for helping us make a useful submodule even more so.  I hope we can address each of the issues you raised here to your satisfaction and in a timely manner.

**Comment by brandondube** [comment]
@stefanv I am sorry that we disagree, but it is my view that someone who's opening comment is:

> No one is forcing you to use SciPy. If you want commercial support, then you can pay for a MATLAB license.
> >You may disagree, but as a user with no stake in Scipy other than my familiarity with it and no malice towards what is on the whole a great effort, that is my opinion and I am entitled to it.

> You fundamentally misunderstand open source software. Fundamentally. SciPy and other open source software isn't written for you. It's written for its contributors, and you're allowed to use it as a side effect. If you're not a contributor, then you are not entitled to decide what it should be called, its development status, or any other aspect of its development or distribution. Perhaps you can convince a contributor to adopt your opinion, but that's unlikely with the tone that you display here.

~ @xoviat 

has no place at this table.  My vote in the matter, as creator of this issue, is ultimate and will change under no circumstance.

I hold no ill-will towards scipy.  I depend on much of it for my work, including the interpolation, delaunay, stats, and optimization submodules.  When I discovered after rewriting my entire program at great personal expense to solve a bug that didn't exist in it but instead existed in scipy, I could have walked away and turned to a different optimization package, or changed languages to something else altogether.  Instead I invested still yet more hours in trying to pinpoint the problem and in writing this issue.  I did this in spite of looking at every issue in this repository related to the optimization submodule and seeing that they were almost unanimously unactioned.

When I wrote this issue, I did so from a perspective of already having read many textual outputs from lbfgsb.f.  I assume that because the timers are included, but broken, and the language is slightly wrong (though I now see iterate vs iteration is somewhat contentious) that the error would lie in that subprogram.  I see now that the error is in lbfgsb.f, or the scipy.optimize.minimize callstack anyway.  I am fully aware that the guess is proved as any sequence and cast/converted to an array.  I would not have expected any of the subroutines (L-BFGS-B, CG, etc) to modify that array in-place instead of creating a new array.  This design choice seems illogical to me, as it is prone to having nasty side effects (such as the callback issue) in exchange for being more performant on the scale of single digit microseconds per iteration.

**Comment by matthew-brett** [comment]
To follow up on what @stefanv is saying - there's a really strong culture in the open-source projects I know, not to block or suppress people, no matter how annoying you may find them or their comments.   Open-source - like Scipy - runs on the fuel of 'everyone is welcome'.   That means it appears to us old-timers as ugly and unhappy if you block someone.   I see you can do it, but we're humbly asking you to be patient; we'll try and help if it gets heated.

**Comment by josef-pkt** [comment]
The scipy organization administrators can block a user, and they should be the only ones that can do it.
https://docs.scipy.org/doc/scipy/reference/dev/conduct/code_of_conduct.html

**Comment by brandondube** [comment]
To me it is not a question of annoying, but gravely offensive.  There is nothing that can be done to get me to lift the ban I have placed on him.

**Comment by matthew-brett** [comment]
@josef-pkt - it looks like a user can block other users from an issue they started.  I don't think we can configure that - at least I can't see that option in a quick look in the interface.

**Comment by josef-pkt** [comment]
It looks like a screwed up permission policy in github. I didn't find any information about this. When I looked for it in the past there was no option for specific issues and I had to completely block a user from the organization.

**Comment by ilayn** [comment]
I think this toxicity will not lead to anything fruitful. I am not happy with watching a frequent contributor is being bashed continuously and banned with the prose of "ultimate owner"not having a "place on the table"  of the issue etc. And I have to repeat, I don't even agree with what xoviat said. 

For a few times you had the chance to straighten out this towards a neutral tone yet you chose to impose a nonexisting authority. So until you work out this hostile behavior this is on the verge of online bullying which you also threatened a user reporting to GitHub.

That for me crosses the line, I don't want to participate in any of this any more and will leave the discussion to the board and other willing SciPy members.

**Comment by ev-br** [comment]
As a matter of policy IMO a user blocking a user is inappropriate. I don't believe that there is ownership of issues.
My recommendation is to ask a blocker to remove the block and refrain from using it in the future. Otherwise the issue should be closed and locked out for further comments.

**Comment by matthew-brett** [comment]
Oh dear - no - I personally think we should reduce the amount of enforcement here, there's already too much.  It's unfortunate we got to this point, and we should really put a rule on this ("don't block from your own issues") somewhere so we can point to it, but I vote for throttling back, and letting this go.  I think it's clear what we prefer.

**Comment by brandondube** [comment]
@ev-br that you would raise the value of @xoviat's voice over improvement of scipy is very saddening to me.

To be clear, I have not done anything special with this issue, merely blocked @xoviat and all of his alts.  If I open another issue on scipy, he will not be able to comment on that either.  His offensive, and abusive tone, is unwelcome and I will not put up with it.

**Comment by rgommers** [comment]
Hi all, can I ask to stop the discussion at this point? It has become unproductive, and has already been reported to the CoC committee at the `scipy-conduct` email adress. The tone of both the initial post and multiple replies to it, by multiple people, have been unhelpful/inappropriate. The code of conduct committee will follow up with the involved people separately.

**Comment by matthew-brett** [comment]
Ouch - that's unfortunate, that it went to the scipy-conduct address before someone pointed to  the relevant section of the CoC doc on this thread.

**Comment by josef-pkt** [comment]
```
This code is not exhaustive or complete. It serves to distill our common understanding of a collaborative, shared environment and goals. Please try to follow this code in spirit as much as in letter, to create a friendly and productive environment that enriches the surrounding community.
```

(no more comments)

**Comment by charris** [comment]
The best policy in these situations is to avoid joining in. A policy have I have just broken, but what is a life without risks :)

**Comment by charris** [comment]
![image](https://user-images.githubusercontent.com/77272/36003571-e65415d2-0ceb-11e8-804e-08b015447f3d.png)

**Comment by brandondube** [comment]
closed; technical issues extracted into #8417 #8418 #8419

**Comment by rgommers** [comment]
*To start with, we apologize for the delay in posting our final summary on this issue. This was mostly not due to lengthly deliberations, we failed to follow up promptly at some point (although the issue with blocking is fairly complex).*

The CoC committee has reviewed this thread (including the deleted
comment), and also talked to both @brandondube and @xoviat privately.
We think there are a few different issues here.

First, @brandondube:
the original ticket blended emotion with technical issues; please
understand that this may upset contributors (as witnessed by the
discussion). This can be avoided in the future by focusing on the issue
at hand, and avoiding statements that relate to the competence of the
developers, the professionality of the work done (this is volunteer
time), etc. We do appreciate your detailed initial report, but if you
want to continue participating on the SciPy issue tracker then you
need to find a way to work productively *with* other contributors, and
to empathize with the volunteers who are trying to improve SciPy while
balancing the needs of many many users, long term maintainability, the
limited time they have available to contribute, etc. We're all on the
same side here. We're closing this thread, technical discussion can
continue on the new issues you opened with more details about your
problem(s). Please keep the above in mind there.

Second, regarding @xoviat's first few posts (one of which has since
been deleted): These were over the line. @xoviat realized this,
contacted us about it privately, and has apologized. (@brandondube, after
talking to you we understand you don't want an apology, but we feel you
ought to know one was offered nevertheless). Having reviewed the posts
in question, we feel an apology is sufficient and consider the matter
settled as far as the SciPy project and the CoC committee is concerned.

Finally, regarding the subsequent interaction between @brandondube and
@xoviat, and the use of Github's blocking feature: this is complicated,
there are a number of things to consider here:
- We do affirm that people do have the right to not have interactions
  with others forced on them whether they want to or not.
  Anything else is raising barriers to victims of harassment.
- Therefore @xoviat's subsequent messages under different usernames are
  problematic, and should not have happened (especially the last one,
  after it was clear to @xoviat that they were blocked).
- *However*, most of those subsequent messages were about finding out what
  was going on (it wasn't clear to anyone that @xoviat had been blocked
  by @brandondube). The issue here is Github's poor UI for the blocking
  feature - we are raising this with Github directly.
- Blocking places us in a dilemma - as SciPy developers, we want all
  contributors to be able to give input on any change; this includes
  @xoviat (he's not being blocked by us from contributing to SciPy as a
  result of this issue).
- Therefore, we recommend strongly that contributors should not block
  each other when an issue first arises. Rather, we would like to see them
  raise the issue with the CoC committee first, which can take appropriate
  action. Only if that doesn't help should a contributor consider blocking
  another contributor.
- In case someone does get blocked, we recommend that they provide any
  technical input they may have via the mailing list (the person who
  blocked them can choose not to participate there, and those posts are
  generally quite well monitored by us all for inappropriate content).

We're locking this issue. Further constructive discussion on or request
for clarification about this statement can be done with the CoC
committee directly.

Signed, the CoC committee (@rgommers, @njsmith, @stefanv)
